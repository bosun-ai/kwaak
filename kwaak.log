OpenTelemetry tracing enabled
>> starting up agent for the first time, this might take a while
