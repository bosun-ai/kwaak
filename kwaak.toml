language = "rust"

tavily_api_key = "env:TAVILY_API_KEY"
github_api_key = "env:GITHUB_TOKEN"
openai_api_key = "env:KWAAK_OPENAI_API_KEY"

tool_executor = "docker"
otel_enabled = true

[commands]
test = "RUST_LOG=kwaak=debug,swiftide=debug RUST_BACKTRACE=1 cargo nextest run --no-fail-fast --color=never"
coverage = "cargo llvm-cov nextest --no-clean --summary-only"
lint_and_fix = "cargo clippy --fix --allow-dirty --allow-staged; cargo fmt"

[git]
owner = "bosun-ai"
repository = "kwaak"
main_branch = "master"

[llm.indexing]
provider = "OpenAI"
prompt_model = "gpt-4o-mini"

[llm.query]
provider = "OpenAI"
prompt_model = "gpt-4o"

[llm.embedding]
provider = "OpenAI"
embedding_model = "text-embedding-3-large"

[docker]
dockerfile = "Dockerfile"
