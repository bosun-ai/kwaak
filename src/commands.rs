use std::sync::Arc;

use anyhow::Result;
use tokio::{sync::mpsc, task};
use uuid::Uuid;

use crate::{
    chat_message::ChatMessage,
    frontend::{App, UIEvent},
    indexing, query,
    repository::Repository,
};

/// Commands represent concrete actions from a user or in the backend
///
/// By default all commands can be triggered from the ui like `/<command>`
#[derive(
    Debug,
    PartialEq,
    Eq,
    strum_macros::EnumString,
    strum_macros::Display,
    strum_macros::IntoStaticStr,
    strum_macros::EnumIs,
    Clone,
)]
#[strum(serialize_all = "snake_case")]
pub enum Command {
    Quit {
        uuid: Uuid,
    },
    ShowConfig {
        uuid: Uuid,
    },
    IndexRepository {
        uuid: Uuid,
    },
    /// Default when no command is provided
    Chat {
        uuid: Uuid,
        message: String,
    },
}

impl Command {
    pub fn uuid(&self) -> Uuid {
        match self {
            Command::Quit { uuid }
            | Command::ShowConfig { uuid }
            | Command::IndexRepository { uuid }
            | Command::Chat { uuid, .. } => *uuid,
        }
    }

    pub fn with_uuid(self, uuid: Uuid) -> Self {
        match self {
            Command::Quit { .. } => Command::Quit { uuid },
            Command::ShowConfig { .. } => Command::ShowConfig { uuid },
            Command::IndexRepository { .. } => Command::IndexRepository { uuid },
            Command::Chat { message, .. } => Command::Chat { uuid, message },
        }
    }
}

/// Commands always flow via the `CommandHandler`
pub struct CommandHandler {
    /// Receives commands
    rx: mpsc::UnboundedReceiver<Command>,
    #[allow(dead_code)]
    /// Sends commands
    tx: mpsc::UnboundedSender<Command>,
    /// Sends `UIEvents` to the connected frontend
    ui_tx: Option<mpsc::UnboundedSender<UIEvent>>,
    /// Repository to interact with
    repository: Arc<Repository>,
}

impl CommandHandler {
    pub fn from_repository(repository: impl Into<Repository>) -> Self {
        let (tx, rx) = mpsc::unbounded_channel();

        CommandHandler {
            rx,
            tx,
            ui_tx: None,
            repository: Arc::new(repository.into()),
        }
    }

    pub fn register_ui(&mut self, app: &mut App) {
        self.ui_tx = Some(app.ui_tx.clone());
        app.command_tx = Some(self.tx.clone());
    }

    pub fn start(mut self) -> tokio::task::JoinHandle<()> {
        task::spawn(async move {
            while let Some(cmd) = self.rx.recv().await {
                let tx = self.ui_tx.clone().expect("Expected a registered ui");
                let repository = Arc::clone(&self.repository);

                tokio::spawn(async move {
                    if let Err(error) =
                        Self::handle_command(&Arc::clone(&repository), &tx, &cmd).await
                    {
                        tracing::error!(?error, %cmd, "Failed to handle command {cmd} with error {error:#}");
                        tx.send(
                            ChatMessage::new_system(format!("Failed to handle command: {error:#}"))
                                .uuid(cmd.uuid())
                                .to_owned()
                                .into(),
                        )
                        .unwrap();
                    }
                });
            }
        })
    }

    /// TODO: Most commands should probably be handled in a tokio task
    /// Maybe generalize tasks to make ui updates easier?
    #[tracing::instrument(skip(repository, tx))]
    async fn handle_command(
        repository: &Repository,
        tx: &mpsc::UnboundedSender<UIEvent>,
        cmd: &Command,
    ) -> Result<()> {
        let now = std::time::Instant::now();
        tracing::warn!("Handling command {cmd}");

        #[allow(clippy::match_wildcard_for_single_variants)]
        match cmd {
            Command::IndexRepository { .. } => indexing::index_repository(repository).await?,
            Command::ShowConfig { uuid } => {
                tx.send(
                    ChatMessage::new_system(toml::to_string_pretty(repository.config())?)
                        .uuid(*uuid)
                        .to_owned()
                        .into(),
                )
                .unwrap();
            }
            Command::Chat { uuid, ref message } => {
                let response = query::query(repository, message).await?;
                tracing::info!(%response, "Chat message received, sending to frontend");
                let response = ChatMessage::new_system(response).uuid(*uuid).to_owned();

                tx.send(response.into()).unwrap();
            }
            // Anything else we forward to the UI
            _ => tx.send(cmd.clone().into()).unwrap(),
        }
        let elapsed = now.elapsed();
        tx.send(
            ChatMessage::new_system(format!(
                "Command {cmd} successful in {} seconds",
                elapsed.as_secs_f64()
            ))
            .uuid(cmd.uuid())
            .into(),
        )
        .unwrap();

        Ok(())
    }
}
