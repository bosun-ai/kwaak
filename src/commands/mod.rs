//! Kwaak uses a command pattern to handle the backend asynchroniously.
mod command;
mod handler;
mod responder;

pub use command::{Command, CommandEvent};
pub use handler::CommandHandler;
pub use responder::{CommandResponse, Responder};

#[cfg(test)]
pub use responder::MockResponder;
