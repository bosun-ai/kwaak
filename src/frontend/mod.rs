mod app;
mod splash;
mod ui_event;
mod ui_input_command;

mod actions;
mod app_command_responder;
/// Different frontend ui modes
mod chat_mode;
mod logs_mode;

/// Let's be very strict about what to export
/// to avoid coupling frontend and the rest
pub use app::App;

#[cfg(debug_assertions)]
pub use app_command_responder::AppCommandResponder;
#[cfg(debug_assertions)]
pub use chat_mode::ui;
#[cfg(debug_assertions)]
pub use ui_event::UIEvent;
#[cfg(debug_assertions)]
pub use ui_input_command::{DiffVariant, UserInputCommand};
