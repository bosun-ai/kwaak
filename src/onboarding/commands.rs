use anyhow::Result;
use serde_json::json;

use crate::onboarding::util::prompt_text;

pub fn command_questions(context: &mut tera::Context) -> Result<()> {
    println!(
        "\nKwaak agents can run tests and use code coverage when coding. Kwaak uses tests as an extra feedback moment for agents"
    );

    let test_command =
        prompt_text("Test command (optional, <esc> to skip)", None).prompt_skippable()?;

    let coverage_command =
        prompt_text("Coverage command (optional, <esc> to skip)", None).prompt_skippable()?;

    context.insert(
        "commands",
        &json!({
            "test": test_command,
            "coverage": coverage_command,
        }),
    );

    Ok(())
}
