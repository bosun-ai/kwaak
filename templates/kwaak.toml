# This is a a prefilled template for `kwaak`
#
# Several assumptions and defaults have been filled in. For proper usage, please customize the values to your needs.
project_name = "{{ project_name }}"
language = "{{ language }}"

## If you are using OpenAI, set the api key here
#openai_api_key = "env:OPENAI_API_KEY"

## Optional: Connect kwaak to github to create PRs, search code, and automatically push to a  remote
#github_api_key = "env:GITHUB_TOKEN"

## Optional: Connect kwaak to tavily to enable it to search the web
#tavily_api_key = "env:TAVILY_API_KEY"

## Commands the agent uses for tools
## Test and coverage are required. Kwaak expects coverage to run the tests as well and echo both the test and coverage results to stdout
## At the moment, the format of the output does not matter.
[commands]
# Example: test = "cargo test --no-fail-fast --color=never"
test = "REQUIRED"
# Example: coverage = "cargo llvm-cov --no-clean --summary-only"
coverage = "REQUIRED"
## Optional: Lint and fix command. This command is run after each completion cycle, before committing the code.
## Recommended to use, as it avoids the LLM getting distracted by linting issues
# Example: lint_and_fix = "cargo clippy --fix --allow-dirty --allow-staged && cargo fmt"

## Git and GitHub configuration
#
## Kwaak can create and update PRs on Github, search github code, and interact with the git repository. This requires a github token.
## If you leave the token empty, kwaak will not create PRs.
[git]
owner = "{{ github.owner }}"
repository = "{{ github.repository }}"
main_branch = "{{ github.main_branch }}"
## The token is expected to have read and write access for git, and the ability to create and update pull requests.
# token = "env:KWAAK_GITHUB_TOKEN"

## Kwaak uses different LLMs for different tasks. As a rule of thumb, tasks that happen often (like indexing, summarizing) require a small, fast model
## and tasks that happen less often (like completion) can use a larger, more accurate model.
#
## You can overwrite the api key per kind of task if needed.
[llm.indexing]
#api_key = "env:KWAAK_OPENAI_API_KEY"
provider = "OpenAI"
prompt_model = "gpt-4o-mini"

[llm.query]
#api_key = "env:KWAAK_OPENAI_API_KEY"
provider = "OpenAI"
prompt_model = "gpt-4o"

[llm.embedding]
#api_key = "env:KWAAK_OPENAI_API_KEY"
provider = "OpenAI"
embedding_model = "text-embedding-3-large"

## Docker configuration
## kwaak requires a Dockerfile for the tool execution environment.
## Besides the dependencies to run the code, there are several additional dependencies:
## - `git` for interacting with the codebase
## - `rg` (ripgrep) for searching the codebase
## - `fd` (fd) for effective file searching
##
## In the future, an executor is planned that does not have these dependencies, but for now, they are required.
##
## If your project already has a Dockerfile and you want to keep it clean, you can specify a different file to use.
[docker]
dockerfile = "Dockerfile"
