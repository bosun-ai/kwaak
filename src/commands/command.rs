use std::sync::Arc;

use derive_builder::Builder;
use uuid::Uuid;

use crate::repository::Repository;

use super::Responder;

/// Commands are the main way to interact with the backend
///
/// By default all commands can be triggered from the ui like `/<command>`
#[derive(
    Debug, strum_macros::Display, strum_macros::IntoStaticStr, strum_macros::EnumIs, Clone,
)]
#[strum(serialize_all = "snake_case")]
pub enum Command {
    /// Cleanly stop the backend
    Quit,

    /// Print the config the config for a repository
    ShowConfig,

    /// Re-index a repository
    IndexRepository,

    /// Stop an agent
    StopAgent,

    /// Chat with an agent
    Chat { message: String },

    /// Get the current changes made by the agent
    Diff,

    /// Execute a command in the context of an agent
    /// and get the output
    Exec { cmd: swiftide::traits::Command },

    /// Retry the last chat with the agent
    /// Will reset history to the point of the last chat, then re-run the chat
    RetryChat,
}

#[derive(Debug, Clone, Builder)]
pub struct CommandEvent {
    command: Command,
    #[builder(default)]
    repository: Option<Arc<Repository>>,
    uuid: Uuid,
    responder: Arc<dyn Responder>,
}

impl CommandEvent {
    #[must_use]
    pub fn builder() -> CommandEventBuilder {
        CommandEventBuilder::default()
    }

    #[must_use]
    pub fn repository(&self) -> Option<&Repository> {
        self.repository.as_deref()
    }

    #[must_use]
    pub fn uuid(&self) -> Uuid {
        self.uuid
    }

    #[must_use]
    pub fn command(&self) -> &Command {
        &self.command
    }

    #[must_use]
    pub fn responder(&self) -> &dyn Responder {
        &self.responder
    }

    #[must_use]
    pub fn clone_responder(&self) -> Arc<dyn Responder> {
        Arc::clone(&self.responder)
    }

    pub fn with_uuid(&mut self, uuid: Uuid) -> &mut Self {
        self.uuid = uuid;
        self
    }

    pub fn with_repository(&mut self, repository: Arc<Repository>) -> &mut Self {
        self.repository = Some(repository);
        self
    }

    pub fn with_maybe_repository(&mut self, repository: Option<Arc<Repository>>) -> &mut Self {
        self.repository = repository;
        self
    }
}

#[cfg(test)]
mod tests {
    use crate::commands::MockResponder;

    use super::*;
    use std::sync::Arc;
    use uuid::Uuid;

    #[test]
    fn test_command_event_builder() {
        let command = Command::Quit;
        let uuid = Uuid::new_v4();
        let responder = Arc::new(MockResponder::new());

        let event = CommandEvent::builder()
            .command(command.clone())
            .uuid(uuid)
            .responder(responder.clone())
            .build()
            .unwrap();

        let dyn_responder = responder as Arc<dyn Responder>;
        assert!(event.command().is_quit());
        assert_eq!(event.uuid(), uuid);
        assert!(Arc::ptr_eq(&event.clone_responder(), &dyn_responder));
    }

    #[test]
    fn test_with_uuid() {
        let command = Command::ShowConfig;
        let uuid = Uuid::new_v4();
        let new_uuid = Uuid::new_v4();
        let responder = Arc::new(MockResponder::new());

        let event = CommandEvent::builder()
            .command(command.clone())
            .uuid(uuid)
            .responder(responder.clone())
            .build()
            .unwrap()
            .with_uuid(new_uuid)
            .to_owned();

        assert_eq!(event.uuid(), new_uuid);
    }
}
