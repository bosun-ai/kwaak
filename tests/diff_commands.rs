use kwaak::commands::Command;
use kwaak::frontend::{ui, DiffVariant, UIEvent, UserInputCommand};
use kwaak::test_utils::{setup_integration, IntegrationContext};
use kwaak::{assert_command_done, git};

/// Tests showing the diff of an agent workspace, and then pulling the diff into a local branch
#[test_log::test(tokio::test(flavor = "multi_thread"))]
async fn test_diff() {
    let IntegrationContext {
        mut app,
        uuid,
        repository,
        mut terminal,
        workdir,

        repository_guard: _repository_guard,
        handler_guard,
    } = setup_integration().await.unwrap();

    // First, let's start a noop agent so an environment is running
    app.dispatch_command(
        uuid,
        Command::Chat {
            message: "hello".to_string(),
        },
    );

    assert_command_done!(app, uuid);

    // The user asks for a diff, it should be empty
    app.send_ui_event(UIEvent::UserInputCommand(
        uuid,
        UserInputCommand::Diff(DiffVariant::Show),
    ));

    assert_command_done!(app, uuid);

    // Now let's add a file and check the diff
    app.dispatch_command(
        uuid,
        Command::Exec {
            cmd: swiftide::traits::Command::write_file("hello.txt", "world"),
        },
    );

    assert_command_done!(app, uuid);

    // now get the diff
    app.send_ui_event(UIEvent::UserInputCommand(
        uuid,
        UserInputCommand::Diff(DiffVariant::Show),
    ));

    assert_command_done!(app, uuid);

    terminal.draw(|f| ui(f, f.area(), &mut app)).unwrap();
    insta::assert_snapshot!(terminal.backend());

    // let's pull the diff
    app.send_ui_event(UIEvent::UserInputCommand(
        uuid,
        UserInputCommand::Diff(DiffVariant::Pull),
    ));

    assert_command_done!(app, uuid);

    let current_branch = git::util::main_branch(&workdir);
    assert_eq!(&current_branch, &repository.config().git.main_branch);

    let branch_name = if let Some(chat) = app.current_chat() {
        chat.branch_name.clone().expect("branch not yet named")
    } else {
        panic!("No current chat");
    };

    // Now let's check out the branch and verify we have the hello.txt
    let output = tokio::process::Command::new("git")
        .arg("checkout")
        .arg(&branch_name)
        .current_dir(&workdir)
        .output()
        .await
        .unwrap();
    dbg!(&output);

    let output = tokio::process::Command::new("git")
        .arg("status")
        .current_dir(&workdir)
        .output()
        .await
        .unwrap();

    dbg!(&output);

    let output = tokio::process::Command::new("git")
        .arg("branch")
        .current_dir(&workdir)
        .output()
        .await
        .unwrap();
    dbg!(&output);

    // And read the file
    let content = std::fs::read_to_string(workdir.join("hello.txt")).unwrap();

    assert_eq!(content, "world\n");

    app.handle_single_event(&UIEvent::ScrollDown).await;

    terminal.draw(|f| ui(f, f.area(), &mut app)).unwrap();
    insta::assert_snapshot!("diff pulled", terminal.backend());

    drop(handler_guard);
}
