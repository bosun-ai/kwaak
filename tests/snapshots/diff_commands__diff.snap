---
source: tests/diff_commands.rs
expression: terminal.backend()
---
"┌──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────↑─────────────Chats─────────────┐"
"│ ✦ Kwek kwek                                                                                                                  █ 󰍩  Kwek                       │"
"│ ℹ Command chat successful in 0 seconds                                                                                       █                               │"
"│ No changes                                                                                                                   █                               │"
"│ ℹ Command diff successful in 0 seconds                                                                                       █                               │"
"│ ℹ Command exec successful in 0 seconds                                                                                       █                               │"
"│ diff --git a/hello.txt b/hello.txt                                                                                           █                               │"
"│ index 5dd01c1..cc628cc 100644                                                                                                █                               │"
"│ --- a/hello.txt                                                                                                              █                               │"
"│ +++ b/hello.txt                                                                                                              █                               │"
"│ @@ -1 +1 @@                                                                                                                  █                               │"
"│ -Hello, world!                                                                                                               █                               │"
"│ \ No newline at end of file                                                                                                  █                               │"
"│ +world                                                                                                                       █                               │"
"│ ℹ Command diff successful in 0 seconds                                                                                       █                               │"
"│                                                                                                                              █                               │"
"│                                                                                                                              █                               │"
"│                                                                                                                              █                               │"
"│                                                                                                                              █─────────Chat commands─────────┤"
"│                                                                                                                              █                               │"
"│                                                                                                                              ║  /next_chat      /help        │"
"│                                                                                                                              ║  /new_chat       /quit        │"
"│                                                                                                                              ║  /delete_chat    /show_config │"
"│                                                                                                                              ║                               │"
"│                                                                                                                              ║                               │"
"│                                                                                                                              ║──────────Keybindings──────────┤"
"│                                                                                                                              ║                               │"
"│                                                                                                                              ║ Page Up/Down - Scroll         │"
"│                                                                                                                              ║ End - Scroll to end           │"
"│                                                                                                                              ↓ Tab - Next chat               │"
"├──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤ ^s - Send message             │"
"│  Send a message to an agent ...                                                                                              │ ^x - Stop agent               │"
"│                                                                                                                              │ ^n - New chat                 │"
"│                                                                                                                              │ ^q - Quit                     │"
"│                                                                                                                              │                               │"
"│                                                                                                                              │                               │"
"│                                                                                                                              │                               │"
"└──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┴───────────────────────────────┘"
"                                                                                                                     kwaak/936da01f-9abd-4d9d-80c7-02af85c822a8 "
"                                                                                                                                                                "
